docs/ci-setup.md

```diff
@@ -0,0 +1,36 @@
+## CI Setup
+
+Setting up appium in an automated environment brings a few challenges with it.
+Any scenario where user interaction is required must be automated or avoided all together.
+For real device set up you should first follow the steps in [this guide](https://github.com/appium/appium-xcuitest-driver/blob/master/docs/real-device-config.md).
+
+### Keychains
+
+One common such scenario is a prompt asking for a keychain to be unlocked in order to sign the WebdriverAgent.
+There are multiple possible solutions for this:
+
+1. Keychains can be set to have no timeout and be unlocked manually once. This can be done using the keychain access application. Sometimes keychains still unlock themselves though and this approach is not recommended.
+2. [It is possible to create a second keychain](https://github.com/appium/appium-xcuitest-driver/blob/master/README.md#real-device-security-settings), which just stores the required certificate to sign the WebdriverAgent.
+The issue with this approach is, that Codesign wants to unlock all listed keychains regardless of the specified keychain, thus leading to a password prompt.
+This can be avoided by setting the default keychain and basically hiding the login keychain at the start of building.
+https://stackoverflow.com/questions/16550594/jenkins-xcode-build-works-codesign-fails
+Not practically when running other build jobs simultaneously.
+3. Stick with the existing keychains as in approach number one, but explicitly call unlock keychain before **each** build. This can be done using fastlane `unlock_keychain` or by using `security unlock-keychain` directly.
+The password can be saved as a CI variable/secret, but this doesn't work for multiple machines with different keychain passwords. Setting the keychain password in each machine's .bashrc as an environment variable does the trick.
```

Comment:
keeping passwords in clear text was never a good idea. I prefer to use vaults for such purpose, where the vault access key is stored in CI and each agent retrieves its credentials using that key

Reply:
Isn't that precisely what the keychain is doing, acting as a vault? Issue is somewhere the access to the vault has to be stored as you said and my suggested way is to use a protected CI variable/secret, when supported by the used CI system.  I don't think adding yet another layer vault on top of the keychain/vault adds more security does it? The alternative is to store it on the machine itself, which obviously renders the whole thing pointless from a security perspective. From my understanding of keychains the unlocking does not happen on a per-process basis but is system wide, which is a massive design flaw in my opinion and the provided security by the keychain concept is questionable either way it gets used. Because of this I just opted to save the keychain password on the machine in plaintext, the keychain gets unlocked system wide for each build job anyway and is essentially permanently open. I know doesn't sound great, but I have been spending a lot of time on getting to work this automated and it's the only reliable way I have figured out, which works with multiple machines and different applications/certificates used in parallel

Reply:
Are you fine with the current description or should I change it? I was thinking maybe I should remove the last sentence in line 19, because then the user can decide whether to follow up with the CI secret recommendation or not himself? I understand that 'advertising' to save a password in plaintext on the machine probably isn't a great idea.

Reply:
we could be more abstract about the topic there. Adding implementation details might make it more confusing. For example, .bashrc is irrelevant for recent macOS versions, which uses zsh as the default CLI interpreter.

Reply:
> we could be more abstract about the topic there. Adding implementation details might make it more confusing. For example, .bashrc is irrelevant for recent macOS versions, which uses zsh as the default CLI interpreter.

You are right. I was only looking from the Gitlab CI perspective and the Gitlab Runner with the shell executor uses bash, but other systems like Jenkins might be different. I changed the description to be more general.

Reply:
yes, works for me